aggregate.go

```diff
@@ -90,32 +144,55 @@ func (a *Aggregator) Aggregate(bundles []Bundle) ([][]Bundle, error) {
 	return aggregatedBundles, nil
 }
 
-func (a *Aggregator) aggregatedBundlesMissVersionBundle(bundles, newGroup []Bundle) bool {
-	// delta is the number of version bundles diverging compared to the aggregated
-	// list of version bundles making up a release. In case delta is greater than
-	// 0, the release misses delta version bundles. In case delta is lower than 0,
-	// the release has delta version bundles too much. The latter should be way
-	// more improbable to happen.
-	var delta int
-	{
-		desireCount := distinctCount(bundles)
-		currentCount := len(newGroup)
-
-		delta = desireCount - currentCount
+func (a *Aggregator) walkTreeAndAddLeaves(n *node, bundles []Bundle) {
+	if len(n.leaves) == 0 {
+		for _, b := range bundles {
+			n.leaves = append(n.leaves, &node{bundle: b})
+		}
+		return
+	}
+
+	for _, leaf := range n.leaves {
+		a.walkTreeAndAddLeaves(leaf, bundles)
 	}
+}
 
-	if delta != 0 {
-		v, err := aggregateReleaseVersion(newGroup)
-		if err != nil {
-			a.logger.Log("level", "error", "message", "failed aggregating release version", "stack", fmt.Sprintf("%#v", err))
-		} else {
-			a.logger.Log("level", "debug", "message", fmt.Sprintf("release misses %d version bundles", delta), "version", v)
+func (a *Aggregator) walkTreeAndAggregate(n *node, bundles []Bundle) [][]Bundle {
+	// If current node is leaf, then return bundle if there are no conflicts.
+	if len(n.leaves) == 0 {
+		// Only aggregate bundle groups that don't have conflicting dependencies.
+		for i, b1 := range bundles {
+			for j, b2 := range bundles {
+				// No need to self-verify.
+				if i == j {
+					continue
+				}
+
+				if a.bundlesConflictWithDependencies(b1, b2) {
+					return [][]Bundle{}
+				}
+
+				if a.bundlesConflictWithDependencies(b2, b1) {
+					return [][]Bundle{}
+				}
+			}
 		}
 
-		return true
+		sort.Sort(SortBundlesByVersion(bundles))
+		sort.Stable(SortBundlesByName(bundles))
+		return [][]Bundle{bundles}
 	}
 
-	return false
+	// In the middle of the tree -> continue walking.
+	aggregates := make([][]Bundle, 0)
+	for _, leaf := range n.leaves {
+		bundlesCopy := make([]Bundle, len(bundles), len(bundles)+1)
+		copy(bundlesCopy, bundles)
+		bundlesCopy = append(bundlesCopy, leaf.bundle)
+		aggregates = append(aggregates, a.walkTreeAndAggregate(leaf, bundlesCopy)...)
```

Comment:
`append` creates a copy so you don't have to call `make` and `copy` explicitly

Reply:
I think only the resulting slice variable (the fat pointer) is copied with `append`. Underlying array may (and often does) be reused and use of `sort` above in leaf would have unwanted side effects in siblings. See https://golang.org/ref/spec#Appending_and_copying_slices
> If the capacity of s is not large enough to fit the additional values, append allocates a new, sufficiently large underlying array that fits both the existing slice elements and the additional values. Otherwise, append re-uses the underlying array.